Auto-sized batching misses out a single record when more than one record

With this test:

```python
    @mock.patch('simple_salesforce.bulk.SFBulkType._add_batch')
    def test_add_autosized_batches(self, add_batch):
        # a sneaky way to get the data back so we can test it
        add_batch.side_effect = lambda job_id, data, operation: data
        sf_bulk_type = SFBulkType(None, None, None, None)
        data = [
            # Expected serialized record size of 13 to 1513. Idea is that earlier record batches are split on record
            # count, whereas later batches are split for hitting the byte limit.
            {'key': 'value' * random.randint(0, i // 50)}
            for i in range(30000)
        ]
        result = sf_bulk_type._add_autosized_batches(data=data, operation="update", job="Job-1")
        reconstructed_data = list(itertools.chain(*result))
        print([len(batch) for batch in result])
        # all data was put in a batch
        self.assertEqual(len(data), len(reconstructed_data))
```

we have the following result:

```
[9999, 10000, 8208, 1792]

Ran 1 test in 1.526s

FAILED (failures=1)

29999 != 30000

Expected :30000
Actual   :29999
<Click to see difference>

Traceback (most recent call last):
  File "/usr/lib64/python3.10/unittest/mock.py", line 1369, in patched
    return func(*newargs, **newkeywargs)
  File ".../simple-salesforce/simple_salesforce/tests/test_bulk.py", line 637, in test_add_autosized_batches
    self.assertEqual(len(data), len(reconstructed_data))
AssertionError: 30000 != 29999
```

It seems that a single record is skipped somewhere, and that the first batch is not quite 10,000 records.

I have a fix and a couple of tests I will make into a PR very shortly...

--- a/simple_salesforce/bulk.py
+++ b/simple_salesforce/bulk.py
@@ -209,10 +209,12 @@ class SFBulkType:
         AND
         number_of_character_limit <= 10000000
 
-        testing for number of characters ensures that file size limit is
-        respected.
+        Documentation on limits can be found at:
+        https://developer.salesforce.com/docs/atlas.en-us.salesforce_app_limits_cheatsheet.meta/salesforce_app_limits_cheatsheet/salesforce_app_limits_platform_bulkapi.htm#ingest_jobs
 
-        this is due to json serialization of multibyte characters.
+        Our JSON serialization uses the default `ensure_ascii=True`, so the
+        character and byte lengths will be the same. Therefore we only need
+        to adhere to a single length limit of 10,000,000 characters.
 
         TODO: In future when simple-salesforce supports bulk api V2
         we should detect api version and set max file size accordingly. V2
@@ -224,28 +226,27 @@ class SFBulkType:
         * Maximum number of characters in a record: 400,000
         * Maximum number of characters in a field: 131,072
         """
-        file_limit = 1024 * 1024 * 10  # 10MB in bytes
-        rec_limit = 10000
-        char_limit = 10000000
+        record_limit = 10_000
+        char_limit = 10_000_000
 
         batches = []
         last_break = 0
-        nrecs, recsize = 0, 0
-        for i, rec in enumerate(data):
-            # 2 is added to account for the enclosing `[]`
-            # and the separator `, ` between records.
-            recsize += len(json.dumps(rec, default=str)) + 2
-            nrecs += len(rec)
+        record_count, char_count = 0, 0
+        for i, record in enumerate(data):
+            # 2 is added to account for the enclosing `[]` for the first record
+            # and the separator `, ` between records for subsequent records.
+            additional_chars = len(json.dumps(record, default=str)) + 2
             if any([
-                recsize >= file_limit,
-                recsize >= char_limit,
-                nrecs >= rec_limit,
-                i >= len(data) - 1
-                ]):
-                i = 1 if len(data) == 1 else i
+                char_count + additional_chars > char_limit,
+                record_count == record_limit
+            ]):
                 batches.append(data[last_break:i])
                 last_break = i
-                nrecs, recsize = 0, 0
+                record_count, char_count = 0, 0
+            char_count += additional_chars
+            record_count += 1
+        if last_break < len(data) - 1:
+            batches.append(data[last_break:])
         return [self._add_batch(job_id=job, data=i,
                                 operation=operation) for i in batches]
 

--- a/simple_salesforce/tests/test_bulk.py
+++ b/simple_salesforce/tests/test_bulk.py
@@ -1,13 +1,19 @@
 """Test for bulk.py"""
 import http.client as http
+import json
+
+import itertools
+import random
 import re
 import unittest
+from unittest import mock
 from unittest.mock import patch
 
 import requests
 import responses
 from simple_salesforce import tests
 from simple_salesforce.api import Salesforce
+from simple_salesforce.bulk import SFBulkType
 from simple_salesforce.exceptions import SalesforceGeneralError
 
 
@@ -612,3 +618,81 @@ class TestSFBulkType(unittest.TestCase):
                             session=session)
         contact = client.bulk.Contact.query_all(data)
         self.assertEqual(self.expected_query, contact)
+
+    @responses.activate
+    @mock.patch('simple_salesforce.bulk.SFBulkType._add_autosized_batches')
+    def test_bulk_operation_auto_batch_size(self, add_autosized_batches):
+        """Test that batch_size="auto" leads to using _add_autosized_batches"""
+        session = requests.Session()
+        client = Salesforce(session_id=tests.SESSION_ID,
+                            instance_url=tests.SERVER_URL,
+                            session=session)
+
+        operation = "update"
+
+        responses.add(
+            responses.POST,
+            re.compile(r'^https://[^/job].*/job$'),
+            body='{"apiVersion": 42.0, "concurrencyMode": "Parallel",'
+            '"contentType": "JSON","id": "Job-1","object": "Contact",'
+            '"operation": "%s","state": "Open"}' % operation,
+            status=http.OK)
+        responses.add(
+            responses.POST,
+            re.compile(r'^https://[^/job].*/job/Job-1$'),
+            body='{"apiVersion" : 42.0, "concurrencyMode" : "Parallel",'
+            '"contentType" : "JSON","id" : "Job-1","object" : "Contact",'
+            '"operation" : "%s","state" : "Closed"}' % operation,
+            status=http.OK
+        )
+        data = [{
+            'AccountId': 'ID-1',
+            'Email': 'contact1@example.com',
+            'FirstName': 'Bob',
+            'LastName': 'x'
+        }]
+        add_autosized_batches.return_value = []
+        client.bulk.Contact._bulk_operation(  # pylint: disable=protected-access
+            operation, data, batch_size="auto"
+        )
+        add_autosized_batches.assert_called_once_with(
+            job='Job-1', data=data, operation=operation
+        )
+
+    @mock.patch('simple_salesforce.bulk.SFBulkType._add_batch')
+    def test_add_autosized_batches(self, add_batch):
+        """Test that _add_autosized_batches batches all records correctly"""
+        # _add_autosized_batches passes the return values from add_batch, so we
+        # can pass the data it was given back so that we can test it
+        add_batch.side_effect = lambda job_id, data, operation: data
+        sf_bulk_type = SFBulkType(None, None, None, None)
+        data = [
+            # Expected serialized record size of 13 to 1513. Idea is that
+            # earlier record batches are split on record count, whereas later
+            # batches are split for hitting the byte limit.
+            {'key': 'value' * random.randint(0, i // 50)}
+            for i in range(30000)
+        ]
+        result = sf_bulk_type._add_autosized_batches(  # pylint: disable=protected-access
+            data=data, operation="update", job="Job-1"
+        )
+        reconstructed_data = list(itertools.chain(*result))
+        # all data was put in a batch
+        self.assertEqual(len(data), len(reconstructed_data))
+        self.assertEqual(data, list(itertools.chain(*result)))
+
+        for i, batch in enumerate(result):
+            record_count = len(batch)
+            size_in_bytes = len(json.dumps(batch))
+            is_last_batch = i == len(result) - 1
+            # Check that all batches are within limits
+            self.assertLessEqual(record_count, 10_000)
+            self.assertLessEqual(size_in_bytes, 10_000_000)
+            # ... and that - except for the last batch - all batches have maxed
+            # out one of the two limits
+            self.assertTrue(
+                is_last_batch or
+                record_count == 10_000 or
+                (size_in_bytes + len(json.dumps(result[i + 1][0])) + 2
+                    > 10_000_000)
+            )

